microerror.go

```diff
@@ -1,43 +1,63 @@
-// Package microerror provides project wide helper functions for a more convenient
-// and efficient error handling.
 package microerror
 
-var (
-	handler Handler = NewErrgoHandler(ErrgoHandlerConfig{
-		CallDepth: 1,
-	})
+import (
+	"fmt"
+	"runtime"
+
+	"errors"
 )
 
-// New returns a new error with the given error message. It is a drop-in
-// replacement for errors.New from the standard library.
-func New(s string) error {
-	return handler.New(s)
+// Cause is here only for backward compatibility purposes and should not be used.
+//
+// NOTE: Use Is instead.
+func Cause(err error) error {
+	e := err
+	for e != nil {
+		err = e
+		e = errors.Unwrap(err)
+	}
+
+	return err
 }
 
-// Newf returns a new error with the given printf-formatted error message.
-func Newf(f string, v ...interface{}) error {
-	return handler.Newf(f, v...)
+func Is(err, target error) bool {
+	return errors.Is(err, target)
 }
 
-// Cause returns the cause of the given error. If the cause of the err can not
-// be found it returns the err itself.
-//
-// Cause is the usual way to diagnose errors that may have been wrapped by Mask
-// or Maskf.
-func Cause(err error) error {
-	return handler.Cause(err)
+func Maskf(err *Error, f string, v ...interface{}) error {
+	annotatedErr := annotatedError{
+		annotation: fmt.Sprintf(f, v...),
+		underlying: err,
+	}
+
+	return mask(annotatedErr)
 }
 
-// Mask is a simple error masker. Masked errors act as tracers within the
-// source code. Inspecting an masked error shows where the error was passed
-// through within the code base. This is gold for debugging and bug hunting.
 func Mask(err error) error {
-	return handler.Mask(err)
+	if err == nil {
+		return nil
+	}
+
+	return mask(err)
+}
+
+func mask(err error) error {
+	_, file, line, _ := runtime.Caller(2)
+
+	return stackedError{
+		stackEntry: StackEntry{
+			File: file,
+			Line: line,
+		},
+		underlying: err,
+	}
 }
 
-// Maskf is like Mask. In addition to that it takes a format string and
-// variadic arguments like fmt.Sprintf. The format string and variadic
-// arguments are used to annotate the given errgo error.
-func Maskf(err error, f string, v ...interface{}) error {
-	return handler.Maskf(err, f, v...)
+// New is here only for backward compatibility purposes and should not be used.
+//
+// NOTE: Use struct initialization literal for Error struct instead.
+func New(kind string) *Error {
```

Comment:
What is the migration path here? We could just go with the big bang and not have `New` anymore.

Reply:
I'm trying to be as backward compatible as possible. See https://github.com/giantswarm/api/pull/858.

For the big bang I have way more to change. Especially that it shouldn't be `*Error` but `Error` and few more things that I'd like to discuss. This is just the base. I prefer to do the big bang once :D

Reply:
Makes sense to me, then let's have one big bang :)

Reply:
It's dropped. Let's do average bang.